pkg/statedb/cell.go

```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package statedb
+
+import (
+	memdb "github.com/hashicorp/go-memdb"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+// This module provides an in-memory database built on top of immutable radix trees
+// (courtesy of github.com/hashicorp/go-memdb). It adapts the go-memdb library for
+// use with Hive by taking the table schemas as a group values from hive and provides
+// typed API (Table[Obj]) for manipulating tables. As the database is based on an
+// immutable data structure, all objects inserted into the database MUST NOT be mutated!
```

Comment:
I'm pretty afraid of this causing us "spook action at a distance" bugs. Without really thinking about the feasability - it'd be very nice to have something akin to the deadlock detecting lock library. Mabye a mutation detecting wrapper for objects stored in the DB, enabled through some debug build trickery? Or static analysis that looks at write txns?

Reply:
I'm pretty afraid too, though since we haven't had that many bugs with cache.Store that is similarly considered immutable I'm not super worried. I added a trivial check in the commit to verify that we're not mutating the same object. We can more fancy mutation detection with something like https://github.com/zimmski/go-mutesting or similar libraries. client-go also has similar code to check that cache.Store objects aren't mutated. I'll leave integrating something like that for a later PR.